### ISSUE TITLE
gateway: unify root identifier resolution errors

### ISSUE THREAD

**Issue by lidel** [ipfs/go-libipfs#185]
> Extracted from https://github.com/ipfs/kubo/pull/9660#pullrequestreview-1310527962 so we dont forget

I was thinking about this again, both from semantic perspective, but also given your feedback on how hairy this gets.
I am leaning towards returning 500s for all invalid root identifiers, just for the sake of consistency and simplicity

@hacdias  would like to hear your thoughts if below makes sense (if you dont have any concerns, feel free to execute on this):

- `/ipfs/this-is-not-a-cid` →  we were thinking about 400 Bad Request, but it may be better/easier to return 500 Server Error, because maybe today we know this CID is not valid, but in 10 years the same software could be wrong? Maybe it is a valid CIDv8 that this old software can't recognize? 400 could be wrong, 500 will always kinda fit.
- `/ipns/this-domain-does-not-exist` → 500 (we dont know if domain does not exist, or was resolution error)
- `/ipns/{cid-libp2p-key-that-cant-be-resolved}` → 500  (if it is a valid key, resolution error, if not, we will fallbck to dnslink, and that also cant be proven negative)

## TLDR

Return HTTP 500 Server Error for all  `/ip[nf]s/id`  failures. 

## TODO

- [x] PR here: https://github.com/ipfs/go-libipfs/pull/187
  - [x] Side-quest:  [`IsErrNotFound` func](https://github.com/ipfs/go-libipfs/blob/00e02499517393b13a22ae293091e861e6f796ca/gateway/handler.go#L581-L586) probably should not have [`routing.ErrNotFound`](https://github.com/libp2p/go-libp2p/blob/1ad0a50be62b1e00388b9d85fc997d66e8910c9d/core/routing/routing.go#L14-L15), because these errors are for /ipns/id, which is we always want to be 500.
- [x] PR in Kubo and tests passing: https://github.com/ipfs/kubo/pull/9660
- [x] PR against ipfs/specs, documenting this behavior https://github.com/ipfs/specs/pull/380

**Comment by hacdias** [comment]
@lidel I completely agree. It's very hard to detect where these errors are coming from without changing a lot of libraries and ensuring we clearly bubble up the errors. I will work on the code to make all requests return HTTP 500 here.

**Comment by iand** [comment]
400 bad request is more appropriate for a CID that the server cannot parse since the server defines what request semantics it supports. The server hasn't failed and the user should not retry the request. A 500 response indicates to the user that they can retry it and expect it to work when the server has recovered.

**Comment by hacdias** [comment]
As I mentioned in Slack, I think we could keep `/ipns` the way it is and only revert to 400 in case there is an invalid CID under the `/ipfs` namespace. That may require some further changes but it is likely possible. However, that's also assuming a CID today and a CID in 10 years will look the same and/or be compatible.

**Comment by lidel** [comment]
@hacdias do you mind filling an issue for this so we tackle it after https://github.com/ipfs/go-libipfs/pull/176 ? (unless you've found an easy fix – maybe detect CID parsing errors in `webError`?)

I hoped this will be cosmetic, but this will be coming back, just like did for the opposite "why 400 on /ipns/, my domain is ok".
(We need to resolve this in both the go-libiofs/gateway library and ipfs/specs).
